v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/IncrementTableCounter.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.templates;
+
+import com.google.cloud.teleport.v2.constants.MetricCounters;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.repackaged.core.org.apache.commons.lang3.StringUtils;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Wait;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IncrementTableCounter extends PTransform<PBegin, PCollection<Void>> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(IncrementTableCounter.class);
+
+  private static final Counter tablesCompleted =
+      Metrics.counter(IncrementTableCounter.class, MetricCounters.TABLES_COMPLETED);
+  private Map<String, Wait.OnSignal<?>> tableWaits;
+
+  private final String shardId;
+
+  public IncrementTableCounter(Map<String, Wait.OnSignal<?>> tableWaits, String shardId) {
+    this.tableWaits = tableWaits;
+    this.shardId = shardId;
+  }
+
+  @Override
+  public PCollection<Void> expand(PBegin input) {
+    List<PCollection<String>> tableNamesPCollections = new ArrayList<>();
+    for (Map.Entry<String, Wait.OnSignal<?>> entry : tableWaits.entrySet()) {
+      String tableName = entry.getKey();
+      Wait.OnSignal<?> wait = entry.getValue();
+      PCollection<String> waitedPCollection =
```

Comment:
This seems a little complex to populate a counter when a table completes. Could you elaborate with a few comments/docs on why this is needed here ?

Reply:
Added, an alternative could potentially be: Run a count transform instead of a pcreate->wait.on. The rest remains the same. But the current approach also allows us to log which table completed.

Reply:
So i think i understand now. The wait on logic has been moved into this this class which is where i was getting confused. 

Ideally, this should be done in the migrate table transform and this counter should just be a simple logger on completion of table. It seems like this class is doing the waiting as well as the logging which is why it is as complex.

I would suggest -
Either move this logic to the migrate Table class.
or
Rename this class to reflect that fact that it is waiting for table completion as bulk of the logic is not the counter.

Reply:
Slight correction, the wait on is not moved here. The parent dependency wait ons are as is. This is an additional wait that needs to be added, which basically says, wait for this table to complete before incrementing counter.

We cannot bypass using this additional waitOn, this is necessary for being able to do a dummy pcreate and increment a counter on completion.

And the issue all this inside the migrateTable class dataflow ends up making a separate counter for every step then. (We end up getting per table singleton counters).

Reply:
Closing as discussed.